src/parser/chunkparser.hpp

```diff
@@ -0,0 +1,94 @@
+#pragma once
+#include <nlohmann/json.hpp>
+#include <array>
+
+namespace docgen {
+namespace details {
+
+// These routines will parse the chunk of text starting at begin
+// until it reaches end and update the json object.
+// When a state changes or the chunk has been fully read, the routine will finish.
+inline void read_routine(const char *&begin, const char *end, nlohmann::json& json)
+{
+    // TODO: implement parsing 
+}
+
+inline void slash_routine(const char *&begin, const char *end, nlohmann::json& json)
+{
+    // TODO: implement parsing 
+}
+
+inline void single_line_routine(const char *&begin, const char *end, nlohmann::json& json)
+{
+    // TODO: implement parsing 
+}
+
+inline void block_routine(const char *&begin, const char *end, nlohmann::json& json)
+{
+    // TODO: implement parsing 
+}
+
+} // namespace details
+
+//////////////////////////////////////////////////////
+// ChunkParser Declaration
+//////////////////////////////////////////////////////
+
+// ChunkParser is a class that will parse a chunk of text into JSON format.
+class ChunkParser
+{
+public:
+
+    // Main function that user will call to parse a chunk of text.
+    // This function will update the json object with information extracted from chunk.
+    void parse(const char *begin, 
+               const char *end,
+               nlohmann::json& json);
```

Comment:
I don't think that unit tests is a good reason to not have `json` be a member of `ChunkParser`—any calls to `parse` on the same `ChunkParser` object with a different `json` would lead to wholly undefined/nonsensical behavior. This seems like the kind of thing you are usually looking to guard against...?

Reply:
It should be user's responsibility to pass in the json object they want to update. Our ChunkParser again is solely responsible for parsing a given input and placing the output somewhere. We cannot enforce that the output be the same for a given object. What if user wants to update different json objects for every chunk of data? Since we're not creating a library, this doesn't matter, but even then I'm still against holding onto a reference as a data member. Dangling reference is mad sketch I don't want to go down that road ever again xD